src/tasks/block.rs

```diff
@@ -109,75 +110,78 @@ impl InProgressBlock {
+        match bundles {
+            Ok(bundles) => {
+                for bundle in bundles {
+                    in_progress.ingest_bundle(bundle);
+                }
+            }
+            Err(e) => {
+                tracing::error!(error = %e, "error polling bundles");
+            }
+        }
+        self.bundle_poller.evict();
+    }
 
-        let handle = tokio::spawn(
+    /// Spawn the block builder task, returning the inbound channel to it, and
+    /// a handle to the running task.
+    pub fn spawn(mut self, outbound: mpsc::UnboundedSender<InProgressBlock>) -> JoinHandle<()> {
+        tokio::spawn(
             async move {
                 loop {
-
-                    select! {
-                        biased;
-                        _ = &mut sleep => {
-                            if !in_progress.is_empty() {
-                                tracing::debug!(txns = in_progress.len(), "sending block to submit task");
-                                let in_progress_block = std::mem::take(&mut in_progress);
-                                if outbound.send(in_progress_block).is_err() {
-                                    tracing::debug!("downstream task gone");
-                                    break
-                                }
-                            }
-
-                            // Reset the sleep timer, as we want to do so when (and only when) our sleep future has elapsed,
-                            // irrespective of whether we have any blocks to build.
-                            sleep.as_mut().reset(tokio::time::Instant::now() + Duration::from_secs(self.incoming_transactions_buffer));
-                        }
-                        tx_resp = tx_inbound.recv() => {
-                            match tx_resp {
-                                Some(tx) => in_progress.ingest_tx(&tx),
-                                None => {
-                                    tracing::debug!("upstream task gone");
-                                    break
-                                }
-                            }
-                        }
-                        bundle_resp = bundle_inbound.recv() => {
-                            match bundle_resp {
-                                Some(bundle) => in_progress.ingest_bundle(bundle),
-                                None => {
-                                    tracing::debug!("upstream task gone");
-                                    break
-                                }
-                            }
+                    // sleep the buffer time
+                    tokio::time::sleep(Duration::from_secs(self.incoming_transactions_buffer))
+                        .await;
+
+                    // Build a block
+                    let mut in_progress = InProgressBlock::default();
+                    self.get_transactions(&mut in_progress).await;
+                    self.get_bundles(&mut in_progress).await;
```

Comment:
an interesting byproduct of this is that bundles will land end of block. nothing to do here now, but something we might wanna think about later

Reply:
yes i think any logic around ordering could be layered in later